ADR for ShortCode Implementation (#69)
cds-snc/notification-adr: records/2024-05-23.scalability.short-code-implementation.md

# Implementation

## Template Categorization

When a client creates a template we want to be able to Categorize it into a certain type of template. We will start with the following Template Categories that were created through research on Notify's current templates.

| Template Category | Example  |
| ----------------- | -------- |
| Status Update | Notice of change in status<br>The progress of a submission |
| <br>Promotional call to action | Surveys<br>General apply now<br>"Learn more" |
| Service-related requests | Submit additional documents<br>Follow up to move a process forward |
| Fulfillment with attachments - email only | "Here’s your permit" |
| <br>Broadcast messages | General information - (not related to transactions such as COVID-19 information)<br>Newsletters<br>Status updates |
| <br>Auto-reply | No-reply messages<br>Acknowledgments<br>Response wait times |
| Verification message | Authentication codes <br>Confirmation of account changes |
| Confirmation / Receipts | Record of transactions<br>Approvals |
| Other | User defined |

### Determining Template Priority

> A new table will be added to the DB for TemplateCategories. The primary key from TemplateCategories will be a foreign key to the Templates Table. All TemplateCategories will be assigned a pre-computed priority.

Reviewer: Fields in the TemplateCategories table useful for UX:
- Category name
- Category description or hint.

# Short Code Implementation

Date: 2025-05-23

## Status

**DRAFT**.

## Context

Notify was awarded a General Purpose Short Code by the Canadian Telecom Regulators. This ADR goes through when an SMS will be sent through a Short Code or Long Code and the decision behind why a certain method will be chosen.

## Implementation Goals
1. Assign SMS templates to send through our generic shortcode.
1. Assign any template to high, normal, or bulk priority automatically (SMS & Email)
1. Design a way to categorize templates
1. Override to force a certain template to go through a certain flow
1. The ability for users to select “Other” and enter their own use case for Template Categories

## Goals to be addressed by another ADR
1. Figure out dynamic and static pools of numbers
1. Decide on using AWS Pinpoint or AWS SNS

> Usage of a pool of codes/ dedicated long codes

Reviewer 2: Assuming that was another numbered point here.
```suggestion
1. Usage of a pool of codes/ dedicated long codes
```

Reviewer 2: There is also a toll free number we use to send the American numbers, which would be an exceptional usage at the moment, but maybe less exceptional if we ever move further to support international SMS sending.

Author: Leaving this as is, to be addressed by the CORE team ADR

## Current Context and Assumptions

1. **We assume that ALL Priority SMS Messages will be sent using the Short Code**

> 1. For the purposes of this ADR, assume SMSs are sent to a send_sms client call. Under the hood the send_sms call might utilize SMS or Pinpoint.

Reviewer 2: ```suggestion
1. For the purposes of this ADR, assume SMSs are sent to a send_sms client call. Under the hood the send_sms call might utilize SNS or Pinpoint.
```

1. Currently, we have 3 priority queues - HIGH, MEDIUM, and LOW. Given a template, your message will reach within a certain SLO if sent through a certain queue

> 1. A template can be of type PRIORITY, NORMAL, and BULK which ***currently*** maps directly to the HIGH, MEDIUM, and LOW queues. 

Reviewer 2: Yeah, that would be nice to change that in the UI and finish the switch to the low/medium/priority terms.

1. The templates refer to both SMS and Email templates, but for shortcodes, this ADR focuses on SMS use cases

# Implementation

## Template Categorization

When a client creates a template we want to be able to Categorize it into a certain type of template. We will start with the following Template Categories that were created through research on Notify's current templates.

> | Template Category | Example  |

Reviewer 3: Should we be defining which of these categories are high priority (short code) here in the ADR?

Author: I will have a discussion with Yael and add this in tomorrow

| ----------------- | -------- |
| Status Update | Notice of change in status<br>The progress of a submission |
| <br>Promotional call to action | Surveys<br>General apply now<br>"Learn more" |
| Service-related requests | Submit additional documents<br>Follow up to move a process forward |
| Fulfillment with attachments - email only | "Here’s your permit" |
| <br>Broadcast messages | General information - (not related to transactions such as COVID-19 information)<br>Newsletters<br>Status updates |
| <br>Auto-reply | No-reply messages<br>Acknowledgments<br>Response wait times |
| Verification message | Authentication codes <br>Confirmation of account changes |
| Confirmation / Receipts | Record of transactions<br>Approvals |

> | Other | User defined |

Reviewer 2: Do we have system status content in SMS? I guess that is mostly around email notifications?

Reviewer 2: Some GCNotify users send email notifications that are basically system status around their own internal system, i.e. _The system failed latest status check_. They integrate with our solution to have a quick email solution. I was wondering if we had such usage for SMS (as we do for emails) because that would be a category of its own.

Reviewer 4: Yes, we do have SMS templates for internal system alerts and monitoring. That category would fall under "Alerts and Monitoring" or some similar wording, and then message type would be SMS.

### Determining Template Priority
A new table will be added to the DB for TemplateCategories. The primary key from TemplateCategories will be a foreign key to the Templates Table. All TemplateCategories will be assigned a pre-computed priority.

<!-- ![Potential ERD](https://raw.githubusercontent.com/cds-snc/notification-adr/records/diagrams/2024-05-23.scalability.short-code-implementation/ERD.jpg?raw=true "Potential ERD") -->

![Potential ERD](../records/diagrams/2024-05-23.scalability.short-code-implementation/ERD.jpg?raw=true "Potential ERD")

```

> process_type is (bulk, normal, priority)

Reviewer 2: Eventually we would move these to `low`, `medium` and `high` (if you do not want to tackle this now). We have a card around this in core team I think.

```

Given the new TemplateCategories, each template will default to a process_type that is set in TemplateCategory. To create manual overrides, an Admin will edit the process_type on the Templates Table. For this to work, we have two options:

> 1. Create a new column in Templates `new_process_type`

Reviewer 2: Sounds confusing and a pitfall for future code maintainers.

> 2. Clear the current process_type in the Templates table to be empty.

Reviewer 2: Yeah I like that.

After the above is implemented,
```
Notification Priority = Template.new_process_type else TemplateCategory.process_type
```
For most Templates, the Template.process_type/new_process_type will be empty. Only if overridden by Admin, will that field have information.

**Use case:**  

> An SMS Template is of type "XYZ" that is of "low" priority. The client reaches out to us about the importance of the messages being sent immediately. We leave the category of the Template as "XYZ", but manually update the Template to "high" priority. This means that the message will be sent using the short-code.

Reviewer 2: Most likely the short code, but there are instances where that might not be the case, i.e. if the destination code is American based, or the service has an associated long code -- or several even, the pools will allow us to do grouping that we could not before.

### Other Template Category

Option 1:  
Let client's manually add a Category for their template, if it doesn't fit in the above usecase. A ticket is created in Freshdesk for this new Template Category. A decision is made about adding a new Category, support can add that Category through an Admin workflow page (similar to Branding).

Option 2:  
Client adds a new Category. We run a classifier to the Templates and determine whether it should be a new Category, or live within the existing Categories (supervised learning). Depending on the decision, a new entry is made in the TemplateCategory table.

Regardless of which option we choose, there is a time when the Category for a Template is undetermined. During this time, we will default the process_type of that Template to a Generic "Low Category (Bulk)" for all of our workflows to stay the same.

#### Example
There is a process_type in TemplateCategories and process_type in Templates.

1. If a client chooses a category, the process_type in Templates will be empty.

> 1. If the client chooses "Other" - while we are determining what TemplateCategory the Template must have, it will default to "Low".

Reviewer 2: Which table does the process type default to *Low*? Do we mean the `TemplateCategory` one here?

Author: correct, we will have a TemplateCategory "Default Low" , and "Other" can be attached to that Category till we decide what the category is

1. If an Admin wants to override the process_type for a given Template, that process_type will be overwritten in the Templates table.

## Existing Templates

In order to categorize existing templates, an offline classifier will be used. Currently Notify Production has 12600 existing templates. There are couple of methods that might be used to classify these templates. The size of the data file is not large, and we can assume that classification might be run locally.

1. Classical Machine Learning Techniques - Naive Bayes, Linear Regression or Random Forest.  

>    We can create a test data set of around 100 templates, and use that to run any of the above models.

Reviewer 5: Do you mean a training data set of around 100 templates? Not sure that'd be enough... and note that these methods would need the training data labelled (ie categories predetermined for the training data)

Reviewer 2: The LLM option would definitely be a lot easier. We can get results out of the box using ChatGPT directly with proper prompt instructions.

Author: discussed irl

1. Leverage GENAI  

>    We can send the data file to a few of the LLM models and run a classifier accordingly. One thing to note is if there are any privacy considerations in sending Template data to a hosted LLM model. 

Reviewer 2: ```suggestion
   We can send the data file to a few of the LLM models and run a classifier accordingly. There are privacy considerations in sending template content to a hosted LLM model that could contain sensitive information from GCNotify's users. We would need to include the policy team to determine the feasibility of this option.
```

1. Local LLM Model  

>    Instead of sending template data to a hosted LLM model, we can run LLama locally and instead use that to help with our classficiation.

Reviewer 2: ```suggestion
   Instead of sending template data to a hosted LLM model, we can run Llama locally and use that to help with our classification.
```

# Short Code Implementation

Date: 2025-05-23

## Status

**DRAFT**.

## Context

Notify was awarded a General Purpose Short Code by the Canadian Telecom Regulators. This ADR goes through when an SMS will be sent through a Short Code or Long Code and the decision behind why a certain method will be chosen.

## Implementation Goals
1. Assign SMS templates to send through our generic shortcode.
1. Assign any template to high, normal, or bulk priority automatically (SMS & Email)
1. Design a way to categorize templates
1. Override to force a certain template to go through a certain flow
1. The ability for users to select “Other” and enter their own use case for Template Categories

## Goals to be addressed by another ADR
1. Figure out dynamic and static pools of numbers

> 1. Decide on using AWS Piinpoint or AWS SNS

Reviewer 5: ```suggestion
1. Decide on using AWS Pinpoint or AWS SNS
```

1. Status updates (eg. notice of change in status, the progress of a submission)
1. Promotional call to action (eg. surveys, general apply now, learn more)
1. Service-related requests (eg. submit additional documents, follow up to move a process forward) 
1. Fulfillment with attachments - email only (eg. here’s your permit)
Broadcast messages (eg. general information, not related to transactions such as COVID-19 information)
1. Auto-reply (eg. no-reply messages, acknowledgments, response wait times)
1. Verification message (eg. authentication codes, confirming an account change)
1. Confirmation / Receipts (eg. record of transaction, approvals)
1. Other

### Determining Template Priority
A new table will be added to the DB for TemplateCategories. The primary key from TemplateCategories will be a foreign key to the Templates Table. All TemplateCategories will be assigned a pre-computed priority.

<!-- ![Potential ERD](https://raw.githubusercontent.com/cds-snc/notification-adr/records/diagrams/2024-05-23.scalability.short-code-implementation/ERD.jpg?raw=true "Potential ERD") -->

> ![Potential ERD](../records/diagrams/2024-05-23.scalability.short-code-implementation/ERD.jpg?raw=true "Potential ERD")

Author: Nope, cause the one in Templates can override the one in TemplateCategories

> ![Potential ERD](../records/diagrams/2024-05-23.scalability.short-code-implementation/ERD.jpg?raw=true "Potential ERD")

Reviewer 2: Good thinking!

```
process_type is (bulk, normal, priority)
```

Given the new TemplateCategories, each template will default to a process_type that is set in TemplateCategory. To create manual overrides (admins will be able to do so for a given template) - Admins will edit the edit the process_type on the Templates Table. For this to work, we have two options:
1. Create a new column in Templates `new_process_type`
2. Clear the current process_type in the Templates table to be empty.

After the above is implemented,
```
Notification Priority = Template.new_process_type else TemplateCategory.process_type
```

> For most Templates, the Template.process_type/new_process_type will be empty. Only if overridden by Admin, will that field have information.

Reviewer 6: If the chosen category determines the process_type and we default it to `low/bulk` in the case of a newly added categories via the `Other` option, then when would `process_type` be empty?

Author: There is a process_type in TemplateCategories and process_type in Templates. 
1. If a client chooses a category, the process_type in Templates will be empty. 
1. If the client chooses "Other" - while we are determining what TemplateCategory the Template must have, it will default to "Low".
1. If an Admin wants to override the process_type for a given Template, that process_type will be overwritten in the Templates table.

Reviewer 6: This makes sense to me since we can override the `process_type`.  I initially thought that the `process_type` for the template would be set with the one associated with the `TemplateCategory`.

Reviewer 2: That could be a third option @whabanks where we could override all `Template.process_type` column to be initially set with the `TemplateCategory.process_type` value... I like though the blank option because it carries extra meaning that we did override the process type and a decision was specifically made for the associated template/service.

**Use case:**  
An SMS Template is of type "XYZ" that is of "low" priority. The client reaches out to us about the importance of the messages being sent immediately. We leave the category of the Template as "XYZ", but manually update the Template to "high" priority. This means that the message will be sent using the short-code.

### Other Template Category

Option 1:  
Let client's manually add a Category for their template, if it doesn't fit in the above usecase. A ticket is created in Freshdesk for this new Template Category. A decision is made about adding a new Category, support can add that Category through an Admin workflow page (similar to Branding).

Option 2:  
Client add's a new Category. We run a classifier to the Templates and determine whether it should be a new Category, or live within the existing Categories (supervised learning). Depending on the decision, a new entry is made in the TemplateCategory table.

> Regardless of Option1 or Option2, while the Category for a Template is undermined - we default the process_type of that Template to a Generic "Low Category (Bulk)" in order for all of our workflows to stay the same.

Reviewer 6: ```suggestion
Regardless of which option we choose, there is a period of time where the Category for a Template is undetermined. During this time, we will default the process_type of that Template to a Generic "Low Category (Bulk)" in order for all of our workflows to stay the same.
```

# Short Code Implementation

Date: 2025-05-23

## Status

**DRAFT**.

## Context

Notify was awarded a General Purpose Short Code by the Canadian Telecom Regulators. This ADR goes through when an SMS will be sent through a Short Code or Long Code and the decision behind why a certain method will be chosen.

## Implementation Goals
1. Assign SMS templates to send through our generic shortcode.
1. Assign any template to high, normal, or bulk priority automatically (SMS & Email)
1. Design a way to categorize templates
1. Override to force a certain template to go through a certain flow
1. The ability for users to select “Other” and enter their own use case for Template Categories

## Goals to be addressed by another ADR
1. Figure out dynamic and static pools of numbers
1. Decide on using AWS Pinpoint or AWS SNS
Usage of a pool of codes/ dedicated long codes

## Current Context and Assumptions

> 1. **We assume that ALL Priority Messages will be sent using the Short Code**

Reviewer 3: ```suggestion
1. **We assume that ALL Priority SMS Messages will be sent using the Short Code**
```

![Potential ERD](../records/diagrams/2024-05-23.scalability.short-code-implementation/ERD.jpg?raw=true "Potential ERD")

```
process_type is (bulk, normal, priority)
```

Given the new TemplateCategories, each template will default to a process_type that is set in TemplateCategory. To create manual overrides, an Admin will edit the process_type on the Templates Table. For this to work, we have two options:
1. Create a new column in Templates `new_process_type`
2. Clear the current process_type in the Templates table to be empty.

After the above is implemented,
```
Notification Priority = Template.new_process_type else TemplateCategory.process_type
```
For most Templates, the Template.process_type/new_process_type will be empty. Only if overridden by Admin, will that field have information.

**Use case:**  
An SMS Template is of type "XYZ" that is of "low" priority. The client reaches out to us about the importance of the messages being sent immediately. We leave the category of the Template as "XYZ", but manually update the Template to "high" priority. This means that the message will be sent using the short-code.

### Other Template Category

Option 1:  

> Let client's manually add a Category for their template, if it doesn't fit in the above usecase. A ticket is created in Freshdesk for this new Template Category. A decision is made about adding a new Category, support can add that Category through an Admin workflow page (similar to Branding).

Reviewer 3: As Yael mentioned today, maybe we don't want to specify this level of detail here about the freshdesk ticket being created, since it may not be the support team carrying out this task.

Author: I think it is just an option - that we might send the "Other" category to freshdesk instead of using a classifier.
